### DISCUSSION TITLE
Can I get the league standings for each scoring period (roto league Basketball)

### DISCUSSION THREAD

**Discussion by imargonis** [cwendt94/espn-api#433] [Q&A]
Hi,
I'd like to make some plots how some stats were changed as the league progressed.

I'd like to start with something simple, just take the standings for each day (if I understand correctly in roto leagues the scoring period is a day) for my league from the start and plot them. Is this data available?

Then I'd also like to take each teams stats per scoring category and plot how they changed from day 1 (the averages)?

Is this possible?

I couldn't understand how to do it.

**Comment by cwendt94** [comment]
Im not very familiar with roto league but you should be able to get the standings and stats each day using the [box scores](https://github.com/cwendt94/espn-api/wiki/League-Class-Basketball) functionality. You can pass the scoring period and matchup period to get all the information about that matchup.

**Reply by imargonis** [reply]
Thank you for the reply. Yes I've tried this.

`league.box_scores(scoring_period=1, matchup_period=1)`

But I get an error in
`box_score.py", line 10, in __init__
    self.home_team = data['home']['teamId']
KeyError: 'home'`

I've tried reconstructing the request URL by looking through the code. I came up with something like this:
`https://fantasy.espn.com/apis/v3/games/fba/seasons/2023/segments/0/leagues/LEAGUE_ID/?view=mMatchupScore&view=mScoreboard&scoringPeriodId=20`

Do you think this is a correct URL? It returns a bunch of data but I am not sure it is correct (but then it is a problem of the API/ESPN and not the library). What I mean by not correct is that although it seems to return some correct stuff like the rosters of that scoring period the scores/points seem to be the latest scores rather of that period (but maybe I am looking at the wrong place).

**Reply by cwendt94** [reply]
That looks to be a bug in the box_score logic. Its assuming there will be a home team which it should not. I just fixed this issue in version 0.29.2

How does roto scoring work because I might need to add a special scoring format for that type. That URL does look correct and If you don't mind could post the response to that call? I can take a look at the data to get a better idea on how the data is formatted and if a new box score type needs to be created.

**Reply by imargonis** [reply]
Thanks for the fix. I update to the latest version and now there is no error. However I do not think there is something in the box_score class that I can use.
 
Ok here I have the response to this API call: https://fantasy.espn.com/apis/v3/games/fba/seasons/2023/segments/0/leagues/LEAGUE_ID/?view=mMatchupScore&view=mScoreboard&scoringPeriodId=20

[response_scoringperiod.json.txt](https://github.com/cwendt94/espn-api/files/10492190/response_scoringperiod.json.txt)

I tried to somewhat anonymize the league/teams.

What I noticed is that the rosters are correct for that scoring period ("rosterForCurrentScoringPeriod"). However the "scoreByStat" actually has the current scores and rank (current rebounds, assists, points etc. of the latest scoring period which is a day). The same is true for "totalPoints", "totalPointsLive" and for "rankCalculatedFinal".

So, for roto scoring, from https://support.espn.com/hc/en-us/articles/360003913972-Scoring-Formats#h_01F0M3FR3EGEG3G9G7KR38W2WQ

> Rotisserie, or "Roto," is the most common way to play fantasy baseball. In this scoring type, teams are ranked from first to last in each statistical category. Points are then awarded according to the order in each category and totaled to determine an overall score and league rank.

The first in each category takes as many points as there are teams in the league (6 in our case) and each subsequent team gets 1 less down to the last team that gets only 1 point.

Important to note here is that teams do not collect points in each scoring period. I might have 30 points today then fall to 28 next day and go up to 31 after that. What matters is the final ranking in each category (i.e. total assists, blocks etc.). In my league we have 8 categories but I think when creating a league you can choose which categories to include (I do not know for sure, I wasn't the one that created the league).

So what I wanted to do is find "historical" data of past scoring periods to see how the rankings have changed over time or like how many steals had each team each day. But I am either having the wrong API call or ESPN does not give out that information (or it is there and I missed it?). Another way would probably be to go through the rosters of each team and add their stats?

**Reply by cwendt94** [reply]
Thanks for getting me that data and explaining the how roto leagues scoring works! 

I will look into the box score/matchup data and see what I can pull from it. 

The rosters of each team should have the players overall stats for the season, but if you want to get the stats from every game you should be able to use the `league.player_info()` [call](https://github.com/cwendt94/espn-api/wiki/League-Class-Basketball) which will return the player and stats from every game they played.

**Reply by cwendt94** [reply]
Here is an initial [PR](https://github.com/cwendt94/espn-api/pull/438) for implementing the roto box score. Still going through the data you provided. 

Would you be able to call this API for your league: https://fantasy.espn.com/apis/v3/games/fba/seasons/2023/segments/0/leagues/LEAGUE_ID?view=mSettings and let me know what the `scoringType` field is. This is the field to determine which box_score class to use when the league is initialized. 

For example head to head category is `H2H_CATEGORY` and head to head points is `H2H_POINTS`

**Reply by imargonis** [reply]
The field's value is "ROTO".

I'll attach the whole response in case you need anything else.
[mSettings.json.txt](https://github.com/cwendt94/espn-api/files/10548005/mSettings.json.txt)

Thank you for doing this.

**Reply by cwendt94** [reply]
Perfect, thank you! If you want to try out this [PR](https://github.com/cwendt94/espn-api/pull/438), should be ready for testing. The response for roto box_score should look like this

```python3
[{
   teams: [{team, wins, ties, losses, points, stats, lineup}]
}]
```
the teams field will be an array of all the teams with their stats for the matchup/scoring period

**Reply by imargonis** [reply]
Sorry for not replying for this long. I did not have the time to test this up until now but I am not sure how to go about testing this (I am really a noob with python).

I have installed espn_api through pip, how would I test the change? I went on to clone the repository locally and checkout the PR but I am not sure how to import the local espn_api instead of the one installed with pip. I've looked into virtualenv etc. but is this the way to go? It seems too complicated. Tried to do relative path import but it does not work.

I understand that it is not on you to teach me python but if you have any tips on how to do this I would be grateful.

**Reply by cwendt94** [reply]
No worries, I have been pretty busy as well!

Already great start cloning the repo and checking out the PR. So if you already have the package installed you should just be able to   use the python cli in the top level repo folder and it will use all of your local changes.

So it would look something like this in your terminal
```
@cwendt94 ➜ /workspaces/espn-api (master) $ python
Python 3.6.15 (default, Dec 21 2021, 12:03:22) 
[GCC 10.2.1 20210110] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from espn_api.basketball import League
>>> league = League(....)
```
